eclair-core/src/main/scala/fr/acinq/eclair/crypto/Sphinx.scala

```diff
@@ -74,298 +48,324 @@ object Sphinx extends Logging {
-    * @param rawPacket      packet received by this node
-    * @return a ParsedPacket(payload, packet, shared secret) object where:
-    *         - payload is the per-hop payload for this node
-    *         - packet is the next packet, to be forwarded using the info that is given in payload (channel id for now)
-    *         - shared secret is the secret we share with the node that sent the packet. We need it to propagate failure
-    *         messages upstream.
-    */
-  def parsePacket(privateKey: PrivateKey, associatedData: ByteVector, rawPacket: ByteVector): Try[ParsedPacket] = Try {
-    require(rawPacket.length == PacketLength, s"onion packet length is ${rawPacket.length}, it should be ${PacketLength}")
-    val packet = Packet.read(rawPacket)
-    val sharedSecret = computeSharedSecret(PublicKey(packet.publicKey), privateKey)
-    val mu = generateKey("mu", sharedSecret)
-    val check = mac(mu, packet.routingInfo ++ associatedData)
-    require(check == packet.hmac, "invalid header mac")
-
-    val rho = generateKey("rho", sharedSecret)
-    val bin = (packet.routingInfo ++ ByteVector.fill(PayloadLength + MacLength)(0)) xor generateStream(rho, PayloadLength + MacLength + MaxHops * (PayloadLength + MacLength))
-    val payload = bin.take(PayloadLength)
-    val hmac = ByteVector32(bin.slice(PayloadLength, PayloadLength + MacLength))
-    val nextRouteInfo = bin.drop(PayloadLength + MacLength)
-
-    val nextPubKey = blind(PublicKey(packet.publicKey), computeblindingFactor(PublicKey(packet.publicKey), sharedSecret))
-
-    ParsedPacket(payload, Packet(Version, nextPubKey.value, hmac, nextRouteInfo), sharedSecret)
-  }
-
-  @tailrec
-  private def extractSharedSecrets(packet: ByteVector, privateKey: PrivateKey, associatedData: ByteVector32, acc: Seq[ByteVector32] = Nil): Try[Seq[ByteVector32]] = {
-    parsePacket(privateKey, associatedData, packet) match {
-      case Success(ParsedPacket(_, nextPacket, sharedSecret)) if nextPacket.isLastPacket => Success(acc :+ sharedSecret)
-      case Success(ParsedPacket(_, nextPacket, sharedSecret)) => extractSharedSecrets(nextPacket.serialize, privateKey, associatedData, acc :+ sharedSecret)
-      case Failure(t) => Failure(t)
+  def peekPayloadLength(payload: ByteVector): Int = {
+    payload.head match {
+      case 0 =>
+        // The 1.0 BOLT spec used 65-bytes frames inside the onion payload.
+        // The first byte of the frame (called `realm`) is set to 0x00, followed by 32 bytes of per-hop data, followed by a 32-bytes mac.
+        65
+      case _ =>
+        // The 1.1 BOLT spec changed the frame format to use variable-length per-hop payloads.
+        // The first bytes contain a varint encoding the length of the payload data (not including the trailing mac).
+        // Since messages are always smaller than 65535 bytes, this varint will either be 1 or 3 bytes long.
+        val dataLength = CommonCodecs.varintoverflow.decode(BitVector(payload.take(3))).require.value.toInt
```

Comment:
How about a rewrite to avoid leaking varint-specific details into this:

```scala
val lengthPrefix = payload.take(3)
val decodeResult = CommonCodecs.varintoverflow.decode(lengthPrefix.bits).require
val dataLength = decodeResult.value
val varintLength = lengthPrefix.length - decodeResult.remainder.length
(varintLength + dataLength + MacLength).toInt
```

Reply:
Good suggestion, I'll update.